sdks/python/apache_beam/transforms/combiners_test.py

```diff
@@ -392,5 +394,92 @@ def test_global_fanout(self):
       assert_that(result, equal_to([49.5]))
 
 
+class LatestTest(unittest.TestCase):
+
+  def test_globally(self):
+    l = [window.GlobalWindows.windowed_value(1, 100),
+         window.GlobalWindows.windowed_value(2, 200),
+         window.GlobalWindows.windowed_value(3, 300)]
+    with TestPipeline() as p:
+      pc = p | Create(l)
+      latest = pc | combine.Latest.Globally()
+      assert_that(latest, equal_to([3]))
+
+  def test_globally_empty(self):
+    l = []
+    with TestPipeline() as p:
+      pc = p | Create(l)
+      latest = pc | combine.Latest.Globally()
+      assert_that(latest, equal_to([None]))
+
+  def test_per_key(self):
+    l = [window.GlobalWindows.windowed_value(('a', 1), 100),
+         window.GlobalWindows.windowed_value(('b', 2), 200),
+         window.GlobalWindows.windowed_value(('a', 3), 300)]
+    with TestPipeline() as p:
+      pc = p | Create(l)
+      latest = pc | combine.Latest.PerKey()
+      assert_that(latest, equal_to([('a', 3), ('b', 2)]))
+
+  def test_per_key_empty(self):
+    l = []
+    with TestPipeline() as p:
+      pc = p | Create(l)
+      latest = pc | combine.Latest.PerKey()
+      assert_that(latest, equal_to([]))
+
+  def test_per_key_type_violation(self):
+    l_dict = [window.GlobalWindows.windowed_value({'a': 1}, 100),
+              window.GlobalWindows.windowed_value({'b': 2}, 100),
+              window.GlobalWindows.windowed_value({'a': 3}, 100)]
+    l_3_tuple = [window.GlobalWindows.windowed_value((1, 2, 3), 100),
+                 window.GlobalWindows.windowed_value((4, 5, 6), 100),
+                 window.GlobalWindows.windowed_value((7, 8, 9), 100)]
+    with self.assertRaises(TypeCheckError):
+      with TestPipeline() as p:
+        pc = p | Create(l_dict)
+        _ = pc | combine.Latest.PerKey()
+
+    with self.assertRaises(TypeCheckError):
+      with TestPipeline() as p:
+        pc = p | Create(l_3_tuple)
+        _ = pc | combine.Latest.PerKey()
+
+
+class LatestCombineFnTest(unittest.TestCase):
+
+  def setUp(self):
+    self.fn = combine.LatestCombineFn()
+
+  def test_create_accumulator(self):
+    accumulator = self.fn.create_accumulator()
+    self.assertEquals(accumulator, (None, window.MIN_TIMESTAMP))
+
+  def test_add_input(self):
+    accumulator = self.fn.create_accumulator()
+    element = (1, 100)
+    new_accumulator = self.fn.add_input(accumulator, element)
+    self.assertEquals(new_accumulator, (1, 100))
+
+  def test_merge_accumulators(self):
+    accumulators = [(1, 100),
+                    (2, 200),
+                    (3, 300)]
+    merged_accumulator = self.fn.merge_accumulators(accumulators)
+    self.assertEquals(merged_accumulator, (3, 300))
+
+  def test_extract_output(self):
+    accumulator = (1, 100)
+    output = self.fn.extract_output(accumulator)
+    self.assertEquals(output, 1)
+
+  def test_input_type_violation(self):
```

Comment:
I think this test should go to `LatestTest` instead of `LatestCombineFnTest`?

Reply:
This test - `test_input_type_violation` is used to test the type imposed on the input to the CombineFn [here](https://github.com/ttanay/beam/blob/e99e4845410e78d77f9e1f82246c851cf38f308c/sdks/python/apache_beam/transforms/combiners.py#L897). 
The test - `test_per_key_type_violation` is used to check that the input is a KV pair in the DoFn - `add_timestamp` of class `PerKey` [here](https://github.com/ttanay/beam/blob/e99e4845410e78d77f9e1f82246c851cf38f308c/sdks/python/apache_beam/transforms/combiners.py#L887). 

I think `test_per_key_type_violation` can be written better to display the same behaviour. I'll make that change.